src/shims/pthread_main_np.h

```diff
@@ -0,0 +1,40 @@
+/*
+ * @APPLE_APACHE_LICENSE_HEADER_START@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * @APPLE_APACHE_LICENSE_HEADER_END@
+ */
+
+#ifndef __DISPATCH_SHIMS_PTHREAD_MAIN_NP__
+#define __DISPATCH_SHIMS_PTHREAD_MAIN_NP__
+
+#if !HAVE_PTHREAD_MAIN_NP
+
+#if __linux__
+#include <unistd.h>
+#include <sys/syscall.h>
+#include <sys/types.h>
+#endif
+
+static inline int
+pthread_main_np()
+{
+#if __linux__
```

Comment:
this is very slow, you should instead base this patch on top of the `__thread` infrastructure that is built built.

Reply:
It is indeed slow. Right now I am building a CFRunLoop based GUI in Swift, and with no input added to the runloop I am getting 0ms per loop on Mac and 17ms  on Linux. With actual input im getting up to 3000ms only on Linux, crazy.

Reply:
Please check https://github.com/PureSwift/Cacao/commit/168b18d583deec82987e9ea0e84d108eac554581